gapis/resolve/memory.go

```diff
@@ -73,7 +73,33 @@ func Memory(ctx context.Context, p *path.Memory) (*service.Memory, error) {
 			interval.Merge(&writes, rng.Window(r).Span(), false)
 		}
 	}
-	api.MutateCmds(ctx, s, nil, cmds[len(cmds)-1])
+	// api.MutateCmds(ctx, s, nil, cmds[len(cmds)-1])
+
+	endCmd := cmds[len(cmds)-1]
+	if syncedApi, ok := endCmd.API().(sync.SynchronizedAPI); ok {
+		requestSubCmdIdx := api.SubCmdIdx(p.After.Indices)
+		syncedApi.MutateSubcommands(ctx, api.CmdID(cmdIdx), endCmd, s,
+			func(s *api.State, subCommandIndex api.SubCmdIdx, cmd api.Cmd) {
+				// Clear the read/write list before grabbing the observations for the
+				// requested subcommand
+				if requestSubCmdIdx.Equals(subCommandIndex) {
+					reads = memory.RangeList{}
+					writes = memory.RangeList{}
+				}
+			}, // preSubCmdCallback
+			func(s *api.State, subCommandIndex api.SubCmdIdx, cmd api.Cmd) {
+				// Set the OnRead and OnWrite to nop after the execution of the
+				// requested subcommand, prevent the reads/writes list being polluted
+				// by following subcommands
+				if requestSubCmdIdx.Equals(subCommandIndex) {
```

Comment:
This is to avoid the memory observation added after the subcommand callbacks to pollute the memory observation for the subcommand.

vkQueueSubmit:
* Read the waiting Semaphores
* execute SubCommand 0
 * collect memory observation for SubCmd 0
* execute SubCommand 1
 * collect memory observation for SubCmd 1
* Read the Semaphores to be signaled

If we don't turn off the OnRead, the memory observation for 'Read the Semaphores to be signaled' will be added to the memory observation of SubCommand 0 and 1.